packages/angular_devkit/build_angular/src/builders/web-test-runner/index.ts

```diff
@@ -0,0 +1,196 @@
+  const buildOutput = await first(
+    buildApplicationInternal(
+      {
+        entryPoints,
+        tsConfig: options.tsConfig,
+        outputPath,
+        aot: false,
+        index: false,
+        outputHashing: OutputHashing.None,
+        optimization: false,
+        externalDependencies: [
+          // Resolved by `@web/test-runner` at runtime with dynamically generated code.
+          '@web/test-runner-core',
+        ],
+        sourceMap: {
+          scripts: true,
+          styles: true,
+          vendor: true,
+        },
+        polyfills,
+      },
+      ctx,
+    ),
+  );
+
+  return buildOutput;
+}
+
+function extractZoneTesting(
+  polyfills: readonly string[],
+): [polyfills: string[], hasZoneTesting: boolean] {
+  const polyfillsWithoutZoneTesting = polyfills.filter(
+    (polyfill) => polyfill !== 'zone.js/testing',
+  );
+  const hasZoneTesting = polyfills.length !== polyfillsWithoutZoneTesting.length;
+
+  return [polyfillsWithoutZoneTesting, hasZoneTesting];
+}
+
+/** Run Web Test Runner on the given directory of bundled JavaScript tests. */
+async function runTests(
+  wtr: typeof WebTestRunner,
+  testDir: string,
+  options: WtrBuilderOptions,
+): Promise<BuilderOutput> {
+  const testPagePath = path.resolve(__dirname, 'test_page.html');
+  const testPage = await fs.readFile(testPagePath, 'utf8');
+
+  const runner = await wtr.startTestRunner({
+    config: {
+      rootDir: testDir,
+      files: [
+        `${testDir}/**/*.js`,
+        `!${testDir}/polyfills.js`,
+        `!${testDir}/chunk-*.js`,
+        `!${testDir}/jasmine.js`,
+        `!${testDir}/jasmine_runner.js`,
+        `!${testDir}/testing.js`, // `zone.js/testing`
+      ],
+      testFramework: {
+        config: {
+          defaultTimeoutInterval: 5_000,
+        },
+      },
+      nodeResolve: true,
+      port: 9876,
+      watch: options.watch ?? false,
+
+      testRunnerHtml: (_testFramework, _config) => testPage,
+    },
+    readCliArgs: false,
+    readFileConfig: false,
+    autoExitProcess: false,
+  });
+  if (!runner) {
+    throw new Error('Failed to start Web Test Runner.');
+  }
+
+  // Wait for the tests to complete and stop the runner.
+  const passed = (await once(runner, 'finished')) as boolean;
+  await runner.stop();
+
+  // No need to return error messages because Web Test Runner already printed them to the console.
+  return { success: passed };
+}
+
+/** Returns the first item yielded by the given generator and cancels the execution. */
+async function first<T>(generator: AsyncIterable<T>): Promise<T> {
+  for await (const value of generator) {
+    return value;
+  }
+
+  throw new Error('Expected generator to emit at least once.');
+}
+
+/** Listens for a single emission of an event and returns the value emitted. */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function once<Map extends Record<string, any>, EventKey extends string & keyof Map>(
```

Comment:
Is this because wtr doesn't expose an actual EventEmitter?

Reply:
Interesting, I wasn't aware of a built-in API for this.

Web Test Runner seems to be using that under the hood, but doesn't expose it. Their `EventEmitter` does not extend or implement Node's `EventEmitter`: https://github.com/modernweb-dev/web/blob/245123e1b92fe0d468fc7ae1e25663f52e775bea/packages/test-runner-core/src/utils/EventEmitter.ts

Those utilities seem nice, and _maybe_ it's compatible enough for them to work, but I'm hesitant to rely on that if we can avoid it.

This is somewhat temporary anyways, once we get to a real watch mode, `once` will go away entirely.